django/db/models/query.py

```diff
@@ -540,18 +541,28 @@ def get_or_create(self, defaults=None, **kwargs):
             params = self._extract_model_params(defaults, **kwargs)
             return self._create_object_from_params(kwargs, params)
 
-    def update_or_create(self, defaults=None, **kwargs):
+    def update_or_create(self, defaults=None, condition=None, **kwargs):
         """
         Look up an object with the given kwargs, updating one with defaults
-        if it exists, otherwise create a new one.
+        if it exists and matches the condition, otherwise create a new one.
         Return a tuple (object, created), where created is a boolean
         specifying whether an object was created.
         """
         defaults = defaults or {}
         self._for_write = True
+
+        # Check if condition is callable and has one input param
+        # otherwise add it to kwargs and pass to model get
+        if not(callable(condition) and len(inspect.signature(condition).parameters) == 1):
+            kwargs['condition'] = condition
+            condition = None
+
         with transaction.atomic(using=self.db):
             try:
                 obj = self.select_for_update().get(**kwargs)
+                if condition and not condition(obj):
+                    return obj, False
+
```

Comment:
Keep what's done in `try` to the minimal expected to raise `self.model.DoesNotExist`. Move this after `except:` block.